File: clients/query_client.py
#!/usr/bin/env python3
"""
Cliente Operador — consulta el servidor vía TCP
Uso: python3 query_client.py <SERVER_HOSTNAME>
"""

import socket
import sys

def resolver_hostname(hostname):
    try:
        ip = socket.gethostbyname(hostname)
        print(f"[DNS] {hostname} → {ip}")
        return ip
    except socket.gaierror:
        print(f"[ERROR] No se pudo resolver {hostname}")
        sys.exit(1)

def enviar_comando(sock, comando):
    sock.sendall((comando + "\n").encode())
    respuesta = sock.recv(4096).decode().strip()
    return respuesta

def main():
    hostname = sys.argv[1] if len(sys.argv) > 1 else 'server.iot-monitor.local'
    port = 9001

    server_ip = resolver_hostname(hostname)

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    sock.connect((server_ip, port))
    print(f"Conectado a {hostname}:{port}\n")

    while True:
        print("\nComandos:")
        print("  1) LIST  — sensores activos")
        print("  2) GET   — datos de un sensor")
        print("  3) STATS — estadísticas")
        print("  4) Salir")
        opcion = input("\nOpción: ").strip()

        if opcion == '1':
            resp = enviar_comando(sock, "LIST|")
            print(f"Respuesta: {resp}")

        elif opcion == '2':
            sid = input("ID del sensor: ").strip()
            resp = enviar_comando(sock, f"GET|{sid}")
            print(f"Respuesta: {resp}")

        elif opcion == '3':
            resp = enviar_comando(sock, "STATS|")
            print(f"Respuesta: {resp}")

        elif opcion == '4':
            break

    sock.close()

if __name__ == '__main__':
    main()

File: clients/QueryClient.java
import java.io.*;
import java.net.*;

public class QueryClient {

    static String resolverHostname(String hostname) throws UnknownHostException {
        InetAddress addr = InetAddress.getByName(hostname);
        System.out.println("[DNS] " + hostname + " → " + addr.getHostAddress());
        return addr.getHostAddress();
    }

    static String enviarComando(PrintWriter out, BufferedReader in, String comando) throws IOException {
        out.println(comando);
        return in.readLine();
    }

    public static void main(String[] args) throws Exception {
        String hostname = args.length > 0 ? args[0] : "server.iot-monitor.local";
        int port = 9001;

        String ip = resolverHostname(hostname);

        Socket socket = new Socket(ip, port);
        PrintWriter out = new PrintWriter(socket.getOutputStream(), true);
        BufferedReader in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
        BufferedReader consola = new BufferedReader(new InputStreamReader(System.in));

        System.out.println("Conectado a " + hostname + ":" + port);

        while (true) {
            System.out.println("\nComandos:");
            System.out.println("  1) LIST  — sensores activos");
            System.out.println("  2) GET   — datos de un sensor");
            System.out.println("  3) STATS — estadísticas");
            System.out.println("  4) Salir");
            System.out.print("\nOpción: ");

            String opcion = consola.readLine().trim();

            if (opcion.equals("1")) {
                String resp = enviarComando(out, in, "LIST|");
                System.out.println("Respuesta: " + resp);

            } else if (opcion.equals("2")) {
                System.out.print("ID del sensor: ");
                String sid = consola.readLine().trim();
                String resp = enviarComando(out, in, "GET|" + sid);
                System.out.println("Respuesta: " + resp);

            } else if (opcion.equals("3")) {
                String resp = enviarComando(out, in, "STATS|");
                System.out.println("Respuesta: " + resp);

            } else if (opcion.equals("4")) {
                break;
            }
        }

        socket.close();
    }
}
